File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/19-TopologicalSort/LeetCode/LeetCode-0207-1.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // Create DAG from prerequisites
        vector<vector<int>> G(numCourses);
        vector<int> inDegree(numCourses, 0);
        
        // Edge: prerequisites[i][1]-->prerequisites[i][0]
        for (int i=0; i<prerequisites.size(); i++){
            G[prerequisites[i][1]].push_back(prerequisites[i][0]);
            inDegree[prerequisites[i][0]]++;
        } // end-for
        
        queue<int> Q; // Queue of nodes with in-degree == 0
        for (int i=0; i<numCourses; i++)
            if (inDegree[i] == 0)
                Q.push(i);
    
        int numNodesTaken = 0;
        while (!Q.empty()){
            auto u = Q.front(); Q.pop();
            numNodesTaken++;
            
            // Reduce In-Degree of all vertices adjacent to u by 1
            for (auto v: G[u]){
                if (--inDegree[v] == 0)
                    Q.push(v);
            } // end-for
        } // end-while
        
        return numNodesTaken == numCourses;
    } // end-canFinish
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/19-TopologicalSort/LeetCode/LeetCode-0207-1.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # Create DAG from prerequisites
        G = []
        for i in range(numCourses): G.append([])
        inDegree = [0]*numCourses
        
        # Edge: prerequisites[i][1]-->prerequisites[i][0]
        for i in range(len(prerequisites)):
            G[prerequisites[i][1]].append(prerequisites[i][0])
            inDegree[prerequisites[i][0]] += 1

        Q = []  # Queue of nodes with in-degree == 0
        for i in range(numCourses):
            if inDegree[i] == 0:
                Q.append(i)
    
        numNodesTaken = 0
        while len(Q) > 0:
            u = Q.pop(0)
            numNodesTaken += 1
            
            # Reduce In-Degree of all vertices adjacent to u by 1
            for v in G[u]:
                inDegree[v] -= 1
                if inDegree[v] == 0:
                    Q.append(v)        
        return numNodesTaken == numCourses
        